ipaserver/plugins/service.py

```diff
@@ -664,6 +674,133 @@ def post_callback(self, ldap, dn, entry_attrs, *keys, **options):
+        excluded = ('ipakrbauthzdata', 'krbprincipalauthind',
+                    'ipakrbrequirespreauth')
+        for arg in self.takes_options:
+            yield arg
+        for arg in super(service_add_smb, self).get_options():
+            check = all([arg not in self.takes_options,
+                         not arg.primary_key,
+                         arg.name not in excluded])
+            if check:
+                yield arg
+
+    def pre_callback(self, ldap, dn, entry_attrs, attrs_list,
+                     *keys, **options):
+        assert isinstance(dn, DN)
+        hostname = keys[0]
+        if len(keys) == 2:
+            netbiosname = keys[1]
+        else:
+            # By default take leftmost label from the host name
+            netbiosname = DNSName.from_text(hostname)[0].decode().upper()
+
+        # SMB service requires existence of the host object
+        # because DCE RPC calls authenticated with GSSAPI are using
+        # host/.. principal by default for validation
+        try:
+            hostresult = self.api.Command['host_show'](hostname)['result']
+        except errors.NotFound:
+            raise errors.NotFound(reason=_(
+                "The host '%s' does not exist to add a service to.") %
+                hostname)
+
+        # We cannot afford the host not being resolvable even for
+        # clustered environments with CTDB because the target name
+        # has to exist even in that case
+        util.verify_host_resolvable(hostname)
+
+        smbaccount = '{name}$'.format(name=netbiosname)
+        smbprincipal = 'cifs/{hostname}'.format(hostname=hostname)
+
+        entry_attrs['krbprincipalname'] = [
+            str(kerberos.Principal(smbprincipal, realm=self.api.env.realm)),
+            str(kerberos.Principal(smbaccount, realm=self.api.env.realm))]
+
+        entry_attrs['krbcanonicalname'] = entry_attrs['krbprincipalname'][0]
+
+        # Rewrite DN using proper rdn and new canonical name because when
+        # LDAPCreate.execute() was called, it set DN to krbcanonicalname=$value
+        dn = DN(('krbprincipalname', entry_attrs['krbcanonicalname']),
+                DN(self.obj.container_dn, api.env.basedn))
+
+        # Enforce ipaKrbPrincipalAlias to aid case-insensitive searches as
+        # krbPrincipalName/krbCanonicalName are case-sensitive in Kerberos
+        # schema
+        entry_attrs['ipakrbprincipalalias'] = entry_attrs['krbcanonicalname']
+
+        for o in ('ipakrbprincipal', 'ipaidobject', 'krbprincipalaux',
+                  'posixaccount'):
+            if o not in entry_attrs['objectclass']:
+                entry_attrs['objectclass'].append(o)
+
+        entry_attrs['uid'] = ['/'.join(
+            kerberos.Principal(smbprincipal).components)]
+        entry_attrs['uid'].append(smbaccount)
+        entry_attrs['cn'] = netbiosname
+        entry_attrs['homeDirectory'] = '/dev/null'
+        entry_attrs['uidNumber'] = '-1'
+        entry_attrs['gidNumber'] = '-1'
```

Comment:
Please change both -1 to baseldap.DNA_MAGIC

Reply:
Done.